runtime-sdk/src/modules/consensus_accounts/mod.rs

```diff
@@ -419,6 +724,82 @@ impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API>
+                if let Some(totals) = reclaims.get(&(ud.epoch, ud.from)) {
+                    *totals
+                } else {
+                    // Fetch consensus height corresponding to the given epoch transition. This
+                    // query may be expensive in case the epoch is far back, but the node is
+                    // guaranteed to have it as it was the state after the last normal round
+                    // (otherwise we would have already processed this epoch).
+                    let height = Consensus::height_for_epoch(ctx, ud.epoch)
+                        .expect("failed to determine height for epoch");
+
+                    // Find the relevant reclaim escrow event.
+                    let totals = ctx
+                        .history()
+                        .consensus_events_at(height, EventKind::Staking)
+                        .expect("failed to fetch historic events")
+                        .iter()
+                        .find_map(|ev| match ev {
+                            consensus::Event::Staking(staking::Event {
+                                escrow:
+                                    Some(staking::EscrowEvent::Reclaim {
+                                        owner,
+                                        escrow,
+                                        amount,
+                                        shares,
+                                    }),
+                                ..
+                            }) if owner == &own_address.into() && escrow == &ud.from.into() => {
+                                Some((amount.try_into().unwrap(), shares.try_into().unwrap()))
+                            }
+                            _ => None,
+                        })
+                        .expect("reclaim event should have been emitted");
+
+                    reclaims.put((ud.epoch, ud.from), totals);
+                    totals
+                };
+
+            // Compute proportion of received amount (shares * total_amount / total_shares).
+            let amount = udi
+                .shares
+                .checked_mul(total_amount)
+                .expect("shares * total_amount should not overflow")
+                .checked_div(total_shares)
+                .expect("total_shares should not be zero");
+            let amount = Consensus::amount_from_consensus(ctx, amount).unwrap();
+            let amount = token::BaseUnits::new(amount, denomination.clone());
+
+            // Mint the given number of tokens.
+            Accounts::mint(ctx, ud.to, &amount).unwrap();
+
+            // Emit undelegation done event.
+            ctx.emit_event(Event::UndelegateDone {
+                from: ud.from,
+                to: ud.to,
+                shares: udi.shares,
+                amount,
+            });
+        }
+    }
 }
 
 impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API> module::InvariantHandler
```

Comment:
Could add an invariant that checks the number of shares the runtime has ecrowed in consensus. Should be `>=`  what is it in its internally tracked delegations, for each consensus account.

Reply:
Added the invariant check.

Reply:
:+1: looks good